pyverbs/providers/efa/efadv.pyx

```diff
@@ -0,0 +1,102 @@
+# SPDX-License-Identifier: (GPL-2.0 OR Linux-OpenIB)
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All rights reserved.
+
+cimport pyverbs.providers.efa.efadv_enums as dve
+cimport pyverbs.providers.efa.libefa as dv
+
+import logging
+
+from pyverbs.base import PyverbsRDMAErrno
+
+
+def dev_cap_to_str(flags):
+    l = {dve.EFADV_DEVICE_ATTR_CAPS_RDMA_READ: 'RDMA Read'}
+    return bitmask_to_str(flags, l)
+
+
+def bitmask_to_str(bits, values):
+    numeric_bits = bits
+    flags = []
+    for k, v in sorted(values.items()):
+        if bits & k:
+            flags.append(v)
+            bits -= k
+    if bits:
+        flags.append(f'??({bits:x})')
+    if not flags:
+        flags.append('None')
+    return ', '.join(flags) + f' ({numeric_bits:x})'
+
+
+cdef class EfaContext(Context):
+    """
+    Represent efa context, which extends Context.
+    """
+    def __init__(self, name=''):
+        """
+        Open an efa device
+        :param name: The RDMA device's name (used by parent class)
+        :return: None
+        """
+        super().__init__(name=name)
+
+    def query_efa_device(self):
+        """
+        Queries the provider for device-specific attributes.
+        :return: An EfaDVDeviceAttr containing the attributes.
+        """
+        dv_attr = EfaDVDeviceAttr()
+        rc = dv.efadv_query_device(self.context, &dv_attr.dv, sizeof(dv_attr.dv))
+        if rc:
+            raise PyverbsRDMAErrno('Failed to query efa device {name}, got {rc}'.
```

Comment:
from the man page, "efadv_query_device" RETURNs the value of errno on failure.
so you should:  "raise PyverbsRDMAError(f'Failed to query efa device {self.name}', rc)".
Raising PyverbsRDMAErrno will get not relevant errno value.

Reply:
Done